xunit-autolabeler-v2/ast_parser/python/source_parsers/direct_invocation.py

```diff
@@ -0,0 +1,53 @@
+# Copyright 2020 Google LLC. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from typing import Any, List
+from ast_parser.lib.constants import IGNORED_METHOD_NAMES
+
+
+def parse(nodes: List[Any], class_name: str) -> List[Any]:
+    """Identify directly-invoked snippet methods in Python
+       files and extract their language-agnostic data
+    Args:
+        nodes: A list of AST nodes obtained via the 'ast' package
+        class_name: the class name of the Python file being parsed
+    Returns:
+        A list of snippet methods (as AST nodes) with added
+        language-agnostic data (in the 'drift' attribute)
+    """
+
+    methods = [x for x in nodes if
+               '.FunctionDef' in str(type(x)) and hasattr(x, 'name')]
+    methods = [x for x in methods if x.name not in IGNORED_METHOD_NAMES]
+
+    # Avoid dupes - don't return already-labelled methods
+    # (this function's result is concat-ed to other method lists!)
+    # --> This works because 'methods' contains global references
+    methods = [x for x in methods if not hasattr(x, 'drift')]
+
+    for m in methods:
+        m.drift = {
```

Comment:
I have a little concern on this. This is just a generic dict. Assuming other language parse has similar code and they can have different schema. Can we have a common schema instead?

Reply:
Are you referring to a common schema across _Python_ parsers, or one across _all language_ parsers?

The latter is hard to enforce programmatically (barring some larger API toolkit like Swagger or an external package like `schema`), except at the meta-parser level (which hasn't been PRed yet). The former could be enforced with something like a Python object.

(If switching from dicts to Python objects would address your concerns here, then I can make that change.)

-------
Also - worth noting: some snippets have different properties than others - e.g. non-HTTP snippets don't have URLs or HTTP methods - so if we use a format that covers _all_ properties, they'd have to be set to some blank value like `''` or None.

Reply:
Python object might work, but if we have single code to generate the drift dict, that might suffice.

Here is the most concerning scenario for me:
1. one day someone decided to use `function_name` instead of `method_name` (or any other change)
2. they forgot to change one place
3. other person bangs their head for debugging

Reply:
Update: chatted offline, agreed to replace these `dict`s with Python objects.

Reply:
Done.